### PR TITLE
CaseOfCase UPLC transformation tests.

### LINKED ISSUE #5958: Verify that CaseOfCase UPLC transformation doesn't make expression stricter than needed
The task is to verify that a UPLC program like this one
```
    case (if b then true else false) of
      true  -> ()
      false -> error ":boom:"
```
doesn't get transformed to 
```
if b
  then ()
  else error ":boom:"
```
where `error` is evaluated strictly.

### PR DESCRIPTION
**Pull request by Unisay**
- [x] Chores: extract `caseOfCase` tests from `Simplifier` module into a dedicated test module.
- [x] Chores: a bit of re-formatting (1st commit)
- [x] A test case that demonstrates how `CaseOfCase` UPLC transformation changes program evaluation result in a presence of `Error`. (2nd commit)
```
Untyped Plutus Core
  CaseOfCase
    CaseOfCase/1:                                  OK
    CaseOfCase/2:                                  OK
    CaseOfCase/3:                                  OK
    Transformation doesn't evaluate error eagerly: FAIL
      untyped-plutus-core/test/Transform/CaseOfCase/Test.hs:92:
      expected: EvaluationSuccess (Constant () (Some (ValueOf DefaultUniUnit ())))
       but got: EvaluationFailure
      Use -p '/CaseOfCase/&&/Transformation doesn'\''t evaluate error eagerly/' to rerun this test only.
```
- [x] Fix for the `CaseOfCase` UPLC transformation to make the test above pass.

Closes #5958

### CODE REVIEW

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/CaseOfCase.hs [hunk 1]
```diff
@@ -27,11 +27,15 @@ import Control.Lens
...
   Case ann scrut alts
-    | ( ite@(Force _ (Builtin _ PLC.IfThenElse))
+    | ( ite@(Force a (Builtin _ PLC.IfThenElse))
         , [cond, (trueAnn, true@Constr{}), (falseAnn, false@Constr{})]
         ) <-
         splitApplication scrut ->
-        mkIterApp
-          ite
-          [cond, (trueAnn, Case ann true alts), (falseAnn, Case ann false alts)]
+        Force a $
```
Comment (Unisay) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/CaseOfCase.hs:34]: I couldn't manufacture an annotation (it is polymorphic), so I used `a` that's been lying around. Most likely its wrong, @effectfully  please tell me about the right way to do it?

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/CaseOfCase.hs [hunk 2]
```diff
@@ -27,11 +27,15 @@ import Control.Lens
...
+          mkIterApp
+            ite
+            [ cond
+            , (trueAnn, Delay trueAnn (Case ann true alts))
+            , (falseAnn, Delay falseAnn (Case ann false alts))
```
Comment (Unisay) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/CaseOfCase.hs:39]: The way annotations are duplicated here feels wrong but idk a better way.

File: plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/Test.hs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+caseOfCase3 = runQuote do
+  b <- freshName "b"
+  x <- freshName "x"
+  xs <- freshName "xs"
+  f <- freshName "f"
+  let ite = Force () (Builtin () PLC.IfThenElse)
+      true = Constr () 0 [Var () x, Var () xs]
+      false = Constr () 1 []
+      altTrue = Var () f
+      altFalse = mkConstant @Integer () 2
```
Comment (effectfully) [inline plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/Test.hs:72]: Oh, @Unisay you can _also_ use `mkConstant`:

```haskell
mkConstant
    :: forall a uni fun term tyname name ann.
       (TermLike term tyname name uni fun, uni `HasTermLevel` a)
    => ann -> a -> term ann
mkConstant ann = constant ann . someValue
```

instead of `someValue`:

```haskell
fromValue :: (HasConstant term, UniOf term `HasTermLevel` a) => a -> term
fromValue = fromValueOf knownUni
```

These are basically equivalent, except the former takes an annotation and the latter is more efficient (inlines better, doesn't worry about annotations).

**Comment by effectfully** [comment]
BTW, if anybody is confused why we even bother optimizing `ifThenElse b (constr <...>) (constr <...>)`, then here's where we produce such expressions:

Screenshot from 2024-05-07 18-12-16

File: plutus-core/untyped-plutus-core/test/Spec.hs [hunk 4]
```diff
@@ -10,29 +11,33 @@ import Evaluation.Builtins (test_builtins)
...
-    , test_freevars
-    , test_parsing
-    , test_debug
-    , test_flat
-    , schnorrVerifyRegressions
-    , evalOrder
-    ]
+  defaultMain $
+    testGroup
+      "Untyped Plutus Core"
```
[the remark this answers is not in the document]
Comment (Unisay) [reply plutus-core/untyped-plutus-core/test/Spec.hs:28]: I like your layout more for aesthetic reasons, but the one I've committed is done by fourmolu and I just go with it, because I like to not care about formatting for a while and then re-format the whole file and make it uniform afterwards.

[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/untyped-plutus-core/test/Spec.hs:28]: Fair enough. I'm surprised that fourmolu does that, I thought it didn't add newlines or something.

File: plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/1.uplc.golden
[the remark this answers is not in the document]
Comment (Unisay) [reply plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/1.uplc.golden:1]: What do you dislike about it? 
(I actually like it)

File: plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/Test.hs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+      alts = V.fromList [altTrue, altFalse]
+  pure $ Case () (mkIterApp ite [((), Var () b), ((), true), ((), false)]) alts
+
+{- |
+
+@
+  case (force ifThenElse) True True False of
```
[the remark this answers is not in the document]
Comment (Unisay) [reply plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/Test.hs:79]: If we want to isolate CaseOfCase transformation we'd better do it explicitly: apply only the CaseOfCase pass.

File: plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/Test.hs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+evaluateUplc
+  :: UPLC.Term Name DefaultUni DefaultFun ()
+  -> EvaluationResult (UPLC.Term Name DefaultUni DefaultFun ())
+evaluateUplc = fst <$> unsafeEvaluateCek noEmitter machineParameters
+ where
+  costModel :: CostModel CekMachineCosts BuiltinCostModel =
+    CostModel defaultCekMachineCosts defaultBuiltinCostModel
+  machineParameters
+    :: MachineParameters CekMachineCosts DefaultFun (CekValue DefaultUni DefaultFun ()) =
+      mkMachineParameters def costModel
```
[the remark this answers is not in the document]
Comment (Unisay) [reply plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/Test.hs:121]: Mostly as a matter of taste 🤷🏼‍♂️ ,

```haskell
  costModel :: CostModel CekMachineCosts BuiltinCostModel =
  costModel = CostModel defaultCekMachineCosts defaultBuiltinCostModel
```
vs.

```haskell
  costModel :: CostModel CekMachineCosts BuiltinCostModel =
    CostModel defaultCekMachineCosts defaultBuiltinCostModel
```

but if I had to justify, I'd say that the latter 
1. occupies less horizontal space and 
2. doesn't repeat `costModel` twice (did you ever make a typo such that name in type sig doesn't match value-level name? "The type signature for `costModel` lacks an accompanying binding...")

Also, we're using two-char wide `::`   (instead of the 1-char wide `∷`) and it doesn't align vertically with the 1-char wide `=`. See how ugly it looks in the first example where `CostModel` is not on the same vertical line.

File: plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/Test.hs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+          ()
+          ( mkIterApp
+              (Force () (Builtin () PLC.IfThenElse))
+              [ ((), mkConstant @Bool () True)
+              , ((), Constr () 0 []) -- True
+              , ((), Constr () 1 []) -- False
+              ]
+          )
+          (V.fromList [mkConstant @() () (), Error ()])
+    simplifiedTerm <- runQuoteT $ simplifyTerm defaultSimplifyOpts def originalTerm
```
[the remark this answers is not in the document]
Comment (Unisay) [reply plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/Test.hs:109]: I'll try to isolate the CaseOfCase transform from other transformations. The less things we test in integration the better it is.

File: plutus-core/untyped-plutus-core/test/Transform/Simplify.hs [hunk 8]
```diff
@@ -272,38 +287,50 @@ forceDelayMultiApply = runQuote $ do
...
-          ( LamAbs () x1 $ LamAbs () x2 $ LamAbs () x3 $ LamAbs () f $
-              Delay () $ mkIterAppNoAnn (Var () f) [Var () x1, Var () x2, Var () x3]
-          )
-          [one, two, three, Var () funcVar]
+            ( LamAbs () x1 $
+                LamAbs () x2 $
+                  LamAbs () x3 $
+                    LamAbs () f $
+                      Delay () $
+                        mkIterAppNoAnn (Var () f) [Var () x1, Var () x2, Var () x3]
```
[the remark this answers is not in the document]
Comment (Unisay) [reply plutus-core/untyped-plutus-core/test/Transform/Simplify.hs:295]: Unfortunately this style doesn't work well with `fourmolu` (and other formatters too). 
I'd have to give up on using the formatter which I personally not ready to do.

File: plutus-core/untyped-plutus-core/test/Transform/Simplify.hs [hunk 9]
```diff
@@ -364,13 +413,13 @@ cse1 = runQuote $ do
...
 cse2 :: Term Name DefaultUni DefaultFun ()
 cse2 = Force () (Force () body)
-  where
-    plus a b = mkIterApp (Builtin () PLC.AddInteger) [((), a), ((), b)]
-    con = mkConstant @Integer ()
-    body = mkIterApp (Builtin () PLC.IfThenElse) [((), cond), ((), true), ((), false)]
-    cond = Apply () (Apply () (Builtin () PLC.LessThanInteger) (con 0)) (con 0)
-    true = Delay () (plus (plus (con 1) (con 2)) (plus (con 1) (con 2)))
-    false = Delay () (plus (con 1) (con 2))
+ where
```
[the remark this answers is not in the document]
Comment (Unisay) [reply plutus-core/untyped-plutus-core/test/Transform/Simplify.hs:416]: ok

**Comment by Unisay** [comment]
- [x] Rebased on `master`.
- [x] Addressed review comments.
- [x] Isolated `caseOfCase` transformation in tests.

File: plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/Test.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/Test.hs:128]: > Also, we're using two-char wide `::`   (instead of the 1-char wide `∷`) and it doesn't align vertically with the 1-char wide `=`. See how ugly it looks in the first example where `CostModel` is not on the same vertical line.

I don't personally find it ugly, I find `=` at the end of a type ugly, but that's all a matter of taste of course.

But the point is that my version is a type _signature_ + a definition, while your version is a type _annotation_ + a definition. I.e. for example this:

```haskell
blah :: forall a. a -> a
blah = f where
    f :: forall b. b -> b
    f = \x -> x
```

type checks, while this:

```haskell
blah :: forall a. a -> a
blah = f where
    f :: forall b. b -> b =
        \x -> x
```

doesn't, giving

```
    • Couldn't match expected type: p0 -> p0
                  with actual type: forall b. b -> b
    • When checking that the pattern signature: forall b. b -> b
```

It's a different kind of thing and you're going against GHC by doing it this way, so don't be surprised if things don't work properly in more complex cases.

Also we use type signatures literally everywhere at the top level, why should non-top level be any different? I also try to use `StandaloneKindSignatures` in as many cases as possible, because the kind of nonsense GHC does at the type level is incomparable to what it does at the term level, so you pretty much have to be very pedantic at the type level, otherwise you're gonna suffer a lot (speaking from personal experience).

All in all, you're going to have an overall more pleasant experience by doing things the way GHC expects you to.

But I don't mind it being written like you did, I'm just not going to comply with any guideline saying I need to do it this way too, should we have such a guideline.

File: plutus-core/untyped-plutus-core/test/Transform/Simplify.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/untyped-plutus-core/test/Transform/Simplify.hs:296]: [Your](https://github.com/IntersectMBO/plutus/pull/5761) own experiments with `fourmolu` suggest otherwise:

Screenshot from 2024-05-08 17-56-06

Why was it fine before, but not now? I only agreed to tolerate fourmolu because it screws up a somewhat small portion of my code. If it's going to screw up every Plutus Core AST I've ever created, then I'm simply not agreeing to that.

**Comment by ana-pantilie** [comment]
> BTW, if anybody is confused why we even bother optimizing `ifThenElse b (constr <...>) (constr <...>)`, then here's where we produce such expressions:
> 
> Screenshot from 2024-05-07 18-12-16

@effectfully why are those functions defined the way they are?

File: plutus-benchmark/nofib/test/9.6/knights10-4x4.budget.golden
Comment (effectfully) [inline plutus-benchmark/nofib/test/9.6/knights10-4x4.budget.golden:1]: One budget golden file change? When there were like 20 in the [original](https://github.com/IntersectMBO/plutus/pull/5337) PR? I don't get it, gonna take a look.

File: plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/Test.hs [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+ where
+  costModel :: CostModel CekMachineCosts BuiltinCostModel =
+    CostModel defaultCekMachineCosts defaultBuiltinCostModel
+  machineParameters
+    :: MachineParameters CekMachineCosts DefaultFun (CekValue DefaultUni DefaultFun ()) =
+      mkMachineParameters def costModel
+
+goldenVsSimplified :: String -> Term Name PLC.DefaultUni PLC.DefaultFun () -> TestTree
+goldenVsSimplified name =
+  goldenVsString name ("untyped-plutus-core/test/Transform/CaseOfCase/" ++ name ++ ".uplc.golden")
```
Comment (effectfully) [inline plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/Test.hs:132]: Don't use `/`, use [`</>`](https://hackage.haskell.org/package/filepath-1.5.2.0/docs/System-FilePath.html#v:-60--47--62-) or `joinPath`. Those handle weird corner cases correctly and are cross-platform.  Also we have the `TestNested` machinery for doing these things, but I see that it wasn't in these tests originally.

File: plutus-benchmark/nofib/test/9.6/knights10-4x4.budget.golden
Comment (effectfully) [reply plutus-benchmark/nofib/test/9.6/knights10-4x4.budget.golden:1]: @Unisay or feel free to take a look yourself why we're not seeing more changes here.

**Comment by effectfully** [comment]
@ana-pantilie 

> Nvm, I see it's because you need to convert from bools to builtin bools. This builtin machinery is so awkward in so many places...

This is very much by design, originally the builtins machinery would return a regular Plutus bool (Scott-encoded back then), but folks wanted me to remove that feature and return built-in bools instead, so I did that in like 2019.

Although now that you mention it, we'll hopefully soon be able to return SOPs from builtins, sounds like we need to come full circle and add builtins returning SOP bools directly.

In any case, for you it's obvious that we should return SOPs and not doing so seems awkward. But for, say, Aiken people it's obvious that we should return constants instead, because they prefer to pretend that SOPs don't exist. It's all subjective.

File: plutus-core/untyped-plutus-core/test/Transform/Simplify.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Unisay) [reply plutus-core/untyped-plutus-core/test/Transform/Simplify.hs:296]: https://github.com/fourmolu/fourmolu/issues/405

File: plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/Test.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/Test.hs:83]: > If we want to isolate CaseOfCase transformation we'd better do it explicitly: apply only the CaseOfCase pass.

I still think it would be better to make sure that _only_ case-of-case applies, even if only for readability purposes. Anyway, I've looked into the code and ironically enough we do rely on case-of-known-constructor not being enabled, because otherwise the case-of-case-optimized expression would optimize further and that would show up in the golden files. So while I think we could do a tiny bit better here, I don't think it's important enough to argue about it.

**Comment by effectfully** [comment]
I've reviewed this once again, I don't think any of my comments are blockers, so let's merge it and do follow-ups:

1. why did only one budget test changed? That's quite weird
2. should we avoid generating `force`-`delay` when branches are pure?

@zliu41 please still review this PR.

**Comment by effectfully** [comment]
Oh, and great work, thank you @Unisay!

**Comment by effectfully** [comment]
@Unisay just made an [issue](https://github.com/IntersectMBO/plutus/issues/5966) for you, not urgent. I decided not to dive into it for now, so hopefully you could.
